mysqld_exporter.go

```diff
@@ -0,0 +1,191 @@
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"strconv"
+	"sync"
+
+	"database/sql"
+	_ "github.com/go-sql-driver/mysql"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	namespace = "mysql"
+)
+
+var (
+	listenAddress = flag.String("web.listen-address", ":9091", "Address to listen on for web interface and telemetry.")
+	metricPath    = flag.String("web.telemetry-path", "/metrics", "Path under which to expose metrics.")
+	configFile    = flag.String("config.file", "mysqld_exporter.conf", "Path to config file.")
+)
+
+type Config struct {
+	Config map[string]string `json:"config"`
+}
+
+type Exporter struct {
+	config                     Config
+	mutex                      sync.RWMutex
+	up                         prometheus.Gauge
+	totalScrapes, errorScrapes prometheus.Counter
+	metrics                    map[string]prometheus.Gauge
+}
+
+func getConfig(file string) (*Config, error) {
+	config := &Config{}
+
+	bytes, err := ioutil.ReadFile(file)
+	if err != nil {
+		return nil, err
+	}
+
+	return config, json.Unmarshal(bytes, &config)
+}
+
+// return new empty exporter
+func NewMySQLExporter(config *Config) *Exporter {
+	return &Exporter{
+		config: *config,
+		up: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "up",
+			Help:      "Was the last scrape of mysqld successful.",
+		}),
+		totalScrapes: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "exporter_total_scrapes",
+			Help:      "Current total mysqld scrapes.",
+		}),
+		errorScrapes: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "exporter_error_scrapes",
+			Help:      "Error mysqld scrapes.",
+		}),
+		metrics: map[string]prometheus.Gauge{},
+	}
+}
+
+func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
+	for _, m := range e.metrics {
+		m.Describe(ch)
+	}
+
+	ch <- e.up.Desc()
+	ch <- e.totalScrapes.Desc()
+	ch <- e.errorScrapes.Desc()
+}
+
+func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+	scrapes := make(chan []string)
+
+	go e.scrape(scrapes)
+
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+	e.setMetrics(scrapes)
+	ch <- e.up
+	ch <- e.totalScrapes
+	ch <- e.errorScrapes
+	e.collectMetrics(ch)
+}
+
+func (e *Exporter) scrape(scrapes chan<- []string) {
+	defer close(scrapes)
+
+	e.totalScrapes.Inc()
+	e.up.Set(0)
```

Comment:
I'm wondering if there is a better way to do this than to set it to 0 then 1 again. If a scrape were to happen in the meantime, wouldn't this flap?